File: web/src/format.js
/** Formatage des dates et libellés, en français. */

const OPTIONS_DATE_HEURE = {
  weekday: 'short', day: '2-digit', month: 'short',
  hour: '2-digit', minute: '2-digit',
}

const OPTIONS_HEURE = { hour: '2-digit', minute: '2-digit' }

export const enDateHeure = (iso) =>
  iso ? new Date(iso).toLocaleString('fr-FR', OPTIONS_DATE_HEURE) : '—'

export const enHeure = (iso) =>
  iso ? new Date(iso).toLocaleString('fr-FR', OPTIONS_HEURE) : '—'

export const enDate = (iso) =>
  iso ? new Date(iso).toLocaleDateString('fr-FR') : '—'

/**
 * « jeu. 12:30 » — indispensable sur les frises d'enregistrement : l'ouverture
 * à H-24 tombe à la même heure que le départ, seul le jour les distingue.
 */
export const enJourHeure = (iso) =>
  iso
    ? new Date(iso).toLocaleString('fr-FR', {
        weekday: 'short', hour: '2-digit', minute: '2-digit',
      })
    : '—'

/**
 * Temps relatif : « à l'instant », « il y a 3 min », « dans 2 h 15 ».
 *
 * Écrit à la main plutôt que délégué à Intl : les libellés courts de
 * l'exploitation ne correspondent pas à ceux que produit la locale, et on
 * veut « dans 2 h 15 » et non « dans 2 heures ».
 */
export function enRelatif(iso) {
  if (!iso) return '—'

  const secondes = Math.round((new Date(iso).getTime() - Date.now()) / 1000)
  const passe = secondes < 0
  const abs = Math.abs(secondes)

  const quantite =
    abs < 45 ? null
    : abs < 3600 ? `${Math.round(abs / 60)} min`
    : abs < 86400 ? dureeHeures(abs)
    : abs < 604800 ? `${Math.round(abs / 86400)} j`
    : `${Math.round(abs / 604800)} sem.`

  if (quantite === null) return "à l'instant"
  return passe ? `il y a ${quantite}` : `dans ${quantite}`
}

/** Durée restante compacte : « 4 h 20 », « 38 min ». */
export function enDuree(iso) {
  if (!iso) return '—'
  const secondes = Math.round((new Date(iso).getTime() - Date.now()) / 1000)
  if (secondes <= 0) return 'échu'
  if (secondes < 3600) return `${Math.max(1, Math.round(secondes / 60))} min`
  return dureeHeures(secondes)
}

function dureeHeures(secondes) {
  const heures = Math.floor(secondes / 3600)
  const minutes = Math.round((secondes % 3600) / 60)
  return `${heures} h ${String(minutes).padStart(2, '0')}`
}

export const LIBELLE_STATUT_VOL = {
  programme: 'Programmé',
  a_lheure: "À l'heure",
  retarde: 'Retardé',
  embarquement: 'Embarquement',
  parti: 'Parti',
  annule: 'Annulé',
}

export const LIBELLE_STATUT_DOSSIER = {
  en_cours: 'En cours',
  enregistre: 'Enregistré',
  annule: 'Annulé',
  embarque: 'Embarqué',
}

export const LIBELLE_FENETRE = {
  pas_encore_ouvert: 'Pas encore ouvert',
  ouvert: 'Ouvert',
  ferme: 'Clôturé',
}

export const VILLES = {
  OUA: 'Ouagadougou', BOY: 'Bobo-Dioulasso', DSS: 'Dakar', ABJ: 'Abidjan',
  BKO: 'Bamako', LFW: 'Lomé', COO: 'Cotonou', NIM: 'Niamey', ACC: 'Accra',
}

export const ville = (code) => VILLES[code] ?? code

/** Regroupe les actions du journal par famille, pour l'icône et la couleur. */
export function familleAction(action = '') {
  if (action.startsWith('guichet.')) return 'guichet'
  if (action.startsWith('vol.')) return 'vol'
  if (action.startsWith('siege.')) return 'siege'
  if (action.startsWith('compte.') || action.startsWith('utilisateur.')) return 'compte'
  if (action.startsWith('dcs.')) return 'dcs'
  return 'enregistrement'
}

/** Initiales d'un nom, pour les pastilles d'avatar. */
export function initiales(nom) {
  if (!nom) return '?'
  return nom.trim().split(/\s+/).slice(0, 2).map((m) => m[0].toUpperCase()).join('')
}

File: mobile/lib/format.dart
/// Formatage des dates et libellés, en français.
///
/// Le pendant Dart de `web/src/format.js` : les deux clients affichent les
/// mêmes chaînes, ce qui évite qu'un passager voie « ven. 04 sept., 12:30 »
/// sur le site et « 04/09/2026 12:30 » dans l'application.
///
/// Les noms de jours et de mois sont écrits ici plutôt que délégués à `intl` :
/// ce dernier exige `initializeDateFormatting('fr_FR')` avant tout usage et
/// lève une exception si on l'oublie. Pour dix mots, la dépendance et son
/// piège d'initialisation ne valent pas le service rendu.
library;

const _jours = ['lun.', 'mar.', 'mer.', 'jeu.', 'ven.', 'sam.', 'dim.'];

const _mois = [
  'janv.', 'févr.', 'mars', 'avr.', 'mai', 'juin',
  'juil.', 'août', 'sept.', 'oct.', 'nov.', 'déc.',
];

DateTime? _lire(String? iso) {
  if (iso == null || iso.isEmpty) return null;
  return DateTime.tryParse(iso)?.toLocal();
}

String _deuxChiffres(int n) => n.toString().padLeft(2, '0');

String _heureDe(DateTime d) => '${_deuxChiffres(d.hour)}:${_deuxChiffres(d.minute)}';

/// « ven. 04 sept., 12:30 »
String enDateHeure(String? iso) {
  final d = _lire(iso);
  if (d == null) return '—';
  return '${_jours[d.weekday - 1]} ${_deuxChiffres(d.day)} ${_mois[d.month - 1]}, ${_heureDe(d)}';
}

/// « 12:30 »
String enHeure(String? iso) {
  final d = _lire(iso);
  return d == null ? '—' : _heureDe(d);
}

/// « 04/09/2026 »
String enDate(String? iso) {
  final d = _lire(iso);
  if (d == null) return '—';
  return '${_deuxChiffres(d.day)}/${_deuxChiffres(d.month)}/${d.year}';
}

/// « jeu. 12:30 » — l'ouverture à H-24 tombe à la même heure que le départ,
/// seul le jour les distingue.
String enJourHeure(String? iso) {
  final d = _lire(iso);
  if (d == null) return '—';
  return '${_jours[d.weekday - 1]} ${_heureDe(d)}';
}

/// Durée restante en clair : « 3 h 16 », « 45 min », « dépassé ».
String enDuree(String? iso) {
  final d = _lire(iso);
  if (d == null) return '—';

  final restant = d.difference(DateTime.now());
  if (restant.isNegative) return 'dépassé';

  final heures = restant.inHours;
  final minutes = restant.inMinutes % 60;

  return heures == 0 ? '$minutes min' : '$heures h ${_deuxChiffres(minutes)}';
}

/// Codes OACI du réseau Air Burkina. Le code brut reste affiché à côté :
/// c'est lui qui figure sur la carte d'embarquement.
const _villes = {
  'OUA': 'Ouagadougou',
  'BOY': 'Bobo-Dioulasso',
  'ABJ': 'Abidjan',
  'DSS': 'Dakar',
  'BKO': 'Bamako',
  'LFW': 'Lomé',
  'NIM': 'Niamey',
  'ACC': 'Accra',
  'COO': 'Cotonou',
  'CKY': 'Conakry',
};

String ville(String? code) => _villes[code] ?? code ?? '—';

const libelleStatutVol = {
  'a_lheure': "À l'heure",
  'retarde': 'Retardé',
  'annule': 'Annulé',
  'embarquement': 'Embarquement',
  'parti': 'Parti',
};

const libelleFenetre = {
  'pas_encore_ouvert': 'Pas encore ouvert',
  'ouvert': 'Ouvert',
  'ferme': 'Fermé',
};

const libelleClasse = {
  'affaires': 'Affaires',
  'economique': 'Économique',
};
